Sources/Tracing/Docs.docc/Guides/ImplementATracer.md

```diff
@@ -0,0 +1,186 @@
+    }
+    set {
+      self[TraceIDKey.self] = newValue
+    }
+  }
+}
+
+var context = DefaultFIXME!!!.topLevel(logger: ...)
+context.baggage.traceID = "4bf92f3577b34da6a3ce929d0e0e4736"
+print(context.baggage.traceID ?? "new trace id")
+```
+
+
+### Extracting & injecting Baggage
+
+When hitting boundaries like an outgoing HTTP request you call out to the [configured instrument(s)](#Bootstrapping-the-Instrumentation-System):
+
+An HTTP client e.g. should inject the given `FIXME!!!` into the HTTP headers of its outbound request:
+
+```swift
+func get(url: String, context: FIXME!!!) {
+  var request = HTTPRequest(url: url)
+  InstrumentationSystem.instrument.inject(
+    context.baggage,
+    into: &request.headers,
+    using: HTTPHeadersInjector()
+  )
+}
+```
+
+On the receiving side, an HTTP server should use the following `Instrument` API to extract the HTTP headers of the given
+`HTTPRequest` into:
+
+```swift
+func handler(request: HTTPRequest, context: FIXME!!!) {
+  InstrumentationSystem.instrument.extract(
+    request.headers,
+    into: &context.baggage,
+    using: HTTPHeadersExtractor()
+  )
+  // ...
+}
+```
+
+> In case your library makes use of the `NIOHTTP1.HTTPHeaders` type we already have an `HTTPHeadersInjector` &
+`HTTPHeadersExtractor` available as part of the `NIOInstrumentation` library.
+
+For your library/framework to be able to carry `FIXME!!!` across asynchronous boundaries, it's crucial that you carry the context throughout your entire call chain in order to avoid dropping metadata.
+
+### Tracing your library
+
+When your library/framework can benefit from tracing, you should make use of it by integrating the `Tracing` library.
+
+In order to work with the tracer [configured by the end-user](#Bootstrapping-the-Instrumentation-System), it adds a property to `InstrumentationSystem` that gives you back a `Tracer`. You can then use that tracer to start `Span`s. In an HTTP client you e.g.
+should start a `Span` when sending the outgoing HTTP request:
+
+```swift
+func get(url: String, context: FIXME!!!) {
+  var request = HTTPRequest(url: url)
+
+  // inject the request headers into the baggage as explained above
+
+  // start a span for the outgoing request
+  let tracer = InstrumentationSystem.tracer
+  var span = tracer.startSpan(named: "HTTP GET", context: context, ofKind: .client)
+
+  // set attributes on the span
+  span.attributes.http.method = "GET"
+  // ...
+
+  self.execute(request).always { _ in
+    // set some more attributes & potentially record an error
+
+    // end the span
+    span.end()
+  }
+}
+```
+
+> ⚠️ Make sure to ALWAYS end spans. Ensure that all paths taken by the code will result in ending the span.
+> Make sure that error cases also set the error attribute and end the span.
+
+> In the above example we used the semantic `http.method` attribute that gets exposed via the
+`TracingOpenTelemetrySupport` library.
+
+## Instrument developers: Creating an instrument
```

Comment:
This is repeated info? L15

Reply:
Hmm, thanks I will reflow this entire guide 👍